check_packages_exist.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/python3
+import argparse
+import gzip
+import pathlib
+from xml.etree import ElementTree
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        prog="check_packages_exist",
+        description="Check to make sure all packages listed in appstream metainfo actually exist in the eopkg index",
+    )
+    parser.add_argument(
+        "appstream_metainfo",
+        action="store",
+        help="Path to the appstream metainfo file to be checked. This must be a gzipped XML file.",
+        type=pathlib.Path,
+    )
+    parser.add_argument(
+        "eopkg_index",
+        action="store",
+        help="Path to the eopkg index file to be validated against. This must be an uncompressed XML file.",
+        type=pathlib.Path,
+    )
+    args = parser.parse_args()
+    appstream_components = get_components_from_appstream(args.appstream_metainfo)
+    eopkg_packages = get_packages_from_solus_index(args.eopkg_index)
+    failed_packages = get_failed_packages(appstream_components, eopkg_packages)
+    if failed_packages:
+        print(
+            "The following packages were present in the provided appstream metainfo file, but not in the eopkg index."
+            " They should be manually deleted from the repository files, and appstream metainfo regenerated."
+        )
+        for package in failed_packages:
+            print(package)
+        exit(1)
+    else:
+        print(
+            "Success! No nonexistent packages were present in the provided appstream metainfo file."
+        )
+        exit(0)
+
+
+def __init__():
+    pass
+
+
+def get_components_from_appstream(xml_path: pathlib.Path) -> iter:
+    """
+    Retrieves a list of components from an AppStream Data XML file. "component" is Appstream language for "Package".
+    :param xml_path: Path to an XML AppStream Data File.
+    :return: An iterable of all components listed in the provided metadata.
+    """
+    appstream_xml = gzip.open(xml_path, "r")
+    tree = ElementTree.parse(appstream_xml)
+    root = tree.getroot()
+    components = [
+        component.find("pkgname").text for component in root.findall("component")
+    ]
+    return components
+
+
+def get_packages_from_solus_index(xml_path: pathlib.Path) -> iter:
```

Comment:
Would it make sense to have a helper/wrapper that points to the canonical unstable url for the compressed xml index and then uncompresses it and passes it to this function?

Maybe even gate it behind a `--fetch-unstable-index` option or something?

Reply:
Well, it would make sense- I'm reluctant to do it, though. Two reasons:

1. The main consumer of this script will be, if I have my way, an Ansible playbook which will be responsible for building and checking appstream data. It could also be used in a GitHub action. Both of these use cases have well-established nice ways of fetching files to be processed, so it seems unnecessary.
2. I hesitate to step outside the python standard library in a simple test script like this. It just seems unnecessary. On the rare occasion you'd be running this by hand, fetching the index also seems pretty easy. The only good way to fetch the index without pulling in additional dependencies would be to shell out to `curl` or `wget`, and while that would work, it seems ugly. And, again, not entirely worthwhile.